services/databaseEncryptionService.ts

```
@@ -100,64 +121,97 @@ export async function initializeEncryption(): Promise<void> {
   }
   return keyCache;
 }
 
-export async function getEncryptionMode(): Promise<EncryptionMode> {
-  try {
-    return await getStoredEncryptionMode();
-  } catch (error) {
-    console.error('Error getting encryption mode:', error);
-    return 'basic';
-  }
-}
-
 export async function reWrapKEK(requireAuth: boolean): Promise<void> {
   if (!keyCache) {
-    throw new Error('Encryption key not initialized.');
+    throw new EncryptionError(ERROR_CODES.UNINITIALIZED_ENCRYPTION, 'Encryption key not initialized.');
+  }
+
+  if (reWrapPromise) {
+    return reWrapPromise;
   }
 
   const newMode: EncryptionMode = requireAuth ? 'protected' : 'basic';
   
-  try {
-    await SecureStore.setItemAsync(
-      SECURE_STORE_KEYS.ENCRYPTION_KEY, 
-      base64.fromByteArray(keyCache), 
-      { requireAuthentication: requireAuth }
-    );
-    await setStoredEncryptionMode(newMode);
-  } catch (error) {
-    throw classifyError(error);
-  }
+  reWrapPromise = (async () => {
+    try {
+      await SecureStore.setItemAsync(
+        SECURE_STORE_KEYS.ENCRYPTION_KEY, 
+        keyCache, 
+        { requireAuthentication: requireAuth }
+      );
+      await setStoredEncryptionMode(newMode);
+    } catch (error) {
+      if (error instanceof EncryptionError) throw error;
+      if (isCancellationError(error)) {
+        throw new EncryptionError(ERROR_CODES.USER_CANCELLED, 'Authentication was cancelled');
+      }
+      throw error;
+    } finally {
+      reWrapPromise = null;
+    }
+  })();
+
+  return reWrapPromise;
 }
 
-export function clearKeyCache(): void {
+export async function clearKeyCache(): Promise<void> {
+  if (initializationPromise) {
+    try {
+      await initializationPromise;
+    } catch {
+      // Swallow errors from initialization since we're clearing anyway
+    }
+  }
+  if (reWrapPromise) {
+    try {
+      await reWrapPromise;
+    } catch {
+      // Swallow errors from reWrap since we're clearing anyway
+    }
+  }
```

What's the point in `await`ing then?

---

@gnarea It's called synchronization. I'm waiting for in-flight operations to complete before clearing the cache to prevent race conditions. 

This is standard cleanup pattern - wait for pending operations to finish, regardless of success/failure, before cleaning up shared state.

The errors are swallowed because we don't care if they succeeded, we just need them to stop running before we clear.